theano/gpuarray/c_code/dnn_fwd.c

```diff
@@ -167,31 +204,17 @@ APPLY_SPECIFIC(conv_fwd)(PyGpuArrayObject *input, PyGpuArrayObject *kerns,
             cuda_exit(c->ctx);
             return 1;
           }
+          prev_algo.algo = algo;
+          // no tensor_op returned from Get()
+          prev_algo.mathType = mathtype = CUDNN_DEFAULT_MATH;
+	  // fprintf(stderr, "(cudnnGetConvolutionForwardAlgorithm: (err:%d), algo: %d\n", err, algo);
       }
-      prev_algo = algo;
-    } else {
-      algo = prev_algo;
+    } else { 
+      algo = (cudnnConvolutionFwdAlgo_t)prev_algo.algo;
+      worksize = prev_algo.wsSize;
+      mathtype = prev_algo.mathType;
     }
-
-    #ifdef DEBUG
-    if (0 != theano_enum_to_string_cudnnConvolutionFwdAlgo_t(algo, algorithm_name))
-        return 1;
-    // NB: This is printed only when algorithm is chosen at runtime.
-    if (reuse_algo)
-        fprintf(stderr, "(reused %s)\n", algorithm_name);
-    else
-        fprintf(stderr, "(using %s)\n", algorithm_name);
-    #endif
-
-    if (params->choose_once) {
-      reuse_algo = 1;
-    } else {
-      for (unsigned int i = 0; i < PyGpuArray_NDIM(input); i++) {
-        prev_img_dims[i] = PyGpuArray_DIM(input, i);
-        prev_kern_dims[i] = PyGpuArray_DIM(kerns, i);
-      }
-    }
-  }
+  } else { /* choose_algo */
```

Comment:
There are tests that fail on Jenkins: http://earlgrey.iro.umontreal.ca:8080/job/Theano_PR/666/

For example, this test is normally managed by fallbacks, but now it fails: http://earlgrey.iro.umontreal.ca:8080/job/Theano_PR/666/testReport/junit/theano.gpuarray.tests/test_dnn/test_batched_conv_small__65537__2__2__2____1__2__2__2___5_/

That's because, for this test, no CUDNN_STATUS_NOT_SUPPORTED is raised, everything seems to goes well but there is a bug in cuDNN that makes it return wrong values.

So, I also think we really need to maintain fallbacks.